src/bin/pg_autoctl/pgsql.h

```diff
@@ -170,6 +170,50 @@ typedef struct NodeAddressArray
 } NodeAddressArray;
 
 
+/*
+ * TimeLineHistoryEntry is taken from Postgres definitions and adapted to
+ * client-size code where we don't have all the necessary infrastruture. In
+ * particular we don't define a XLogRecPtr data type nor do we define a
+ * TimeLineID data type.
+ *
+ * Zero is used indicate an invalid pointer. Bootstrap skips the first possible
+ * WAL segment, initializing the first WAL page at WAL segment size, so no XLOG
+ * record can begin at zero.
+ */
+#define InvalidXLogRecPtr 0
+#define XLogRecPtrIsInvalid(r) ((r) == InvalidXLogRecPtr)
+
+#define PG_AUTOCTL_MAX_TIMELINES 1024
```

Comment:
It seems dangerous to put a limit on this, since the amount of timelines could reasonably be very high. It might be better to use a `malloc` instead.

Reply:
I would like to avoid `malloc` and then having to deal with who's responsible for calling `free` as much as possible, here we don't have to think about the lifetime of the “cache”. It's trivial and quite cheap to enlarge that array even tenfold, if you think we need to do that?

At the moment the array (given 64 bits alignment) is going to be 128kB of memory...